### ISSUE TITLE
Bug: <short description of the issue>

### ISSUE THREAD

**Issue by brettz9** [typescript-eslint/typescript-eslint#5730]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8CwAoa6AOBXARiAlsNAE4CmA5vhAC6nEDCWNA9gLYAKhA1lhgBSp0Q4MxAEI+ZgDsAYlilEAvND4BKaIoB80asSykANIPSqAXNABuzfABNEAX1SOUQA

### Repro Code

```typescript
const parser = require('@typescript-eslint/parser');

parser.parse(`class A {
   public registerCustomPickup(
     collisionFunc = () => true,
   ): void {}
 }
`);
```

### ESLint Config

```javascript
N/A
```

### tsconfig

```jsonc
N/A
```

### Expected Result

Not to get an error

### Actual Result

Got this error:

```
/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/FunctionScope.js:28
        const bodyStart = (_b = (_a = this.block.body) === null || _a === void 0 ? void 0 : _a.range[0]) !== null && _b !== void 0 ? _b : -1;
                                                                                                    ^

TypeError: Cannot read properties of undefined (reading '0')
    at FunctionScope.isValidResolution (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/FunctionScope.js:28:101)
    at resolve (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/ScopeBase.js:183:27)
    at /Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/ScopeBase.js:196:18
    at /Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/ScopeBase.js:285:43
    at Array.forEach (<anonymous>)
    at FunctionScope.close (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/scope/ScopeBase.js:285:28)
    at Referencer.close (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/referencer/Referencer.js:51:66)
    at ClassVisitor.visitMethodFunction (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/referencer/ClassVisitor.js:181:69)
    at ClassVisitor.visitMethod (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/referencer/ClassVisitor.js:207:18)
    at ClassVisitor.MethodDefinition (/Users/brett/eslint-plugin-jsdoc/node_modules/.pnpm/@typescript-eslint+scope-manager@5.39.0/node_modules/@typescript-eslint/scope-manager/dist/referencer/ClassVisitor.js:267:14)
```

### Additional Info

Removing the default `() => true` avoids the error

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `5.39.0` |
| `node`                                 | `16.14.2` |

**Comment by bradzacher** [comment]
Our `parser` package isn't intended to be used programmatically - it's a package for use by ESLint. It expects to be called exactly how ESLint calls it.

If you're looking for a parser to use programmatically - you should consider using our `typescript-estree` package instead.

**Comment by brettz9** [comment]
My purpose was to give a simpler demo, as I would have expected that `parse` or at least `parseForESLint` (which gives the same error) would not err here.

My issue is that if a decorator is added above the method (`@Extended`), in ESLint, I get a firing of `ArrowFunctionExpression` but not without it. I wouldn't expect it to be different, but I figured that perhaps that could be an ESLint issue so thought the programmatic approach might be better to describe the issue.

**Comment by bradzacher** [comment]
The issue is by providing a different repro, you've not reproduced your bug at all.

If you've got a bug with a specific selector like that - best way to report it is using a repro repo which shows the incorrect behaviour so that we can exactly reproduce the problem you want us to look into.
